File: services/backend/src/TraderPro.Domain/Common/Measurements/WeightProcessor.cs
using System.Globalization;
using System.Numerics;

namespace TraderPro.Domain.Common.Measurements;

public static class WeightProcessor
{
    private const int StorageDecimalPlaces = 6;
    private const int MaximumIntegerDigits = 14;
    private const int MaximumFractionDigits = 6;
    private static readonly BigInteger StorageScale =
        BigInteger.Pow(10, StorageDecimalPlaces);

    public static WeightProcessingResult Process(
        string? rawWeightKg,
        int decimalPlaces,
        string? method)
    {
        return Process(
            rawWeightKg,
            WeightProcessingPolicy.Parse(decimalPlaces, method));
    }

    public static WeightProcessingResult Process(
        string? rawWeightKg,
        WeightProcessingPolicy policy)
    {
        ArgumentNullException.ThrowIfNull(policy);

        var rawScaledWeight = ParseRawWeight(rawWeightKg);
        var quantum = BigInteger.Pow(
            10,
            StorageDecimalPlaces - policy.DecimalPlaces);
        var units = BigInteger.DivRem(
            rawScaledWeight,
            quantum,
            out var remainder);

        units = policy.Method switch
        {
            WeightProcessingMethod.Standard
                when remainder * 2 >= quantum => units + 1,
            WeightProcessingMethod.Ceiling
                when !remainder.IsZero => units + 1,
            WeightProcessingMethod.Standard or
                WeightProcessingMethod.Floor or
                WeightProcessingMethod.Ceiling => units,
            _ => throw new WeightProcessingException(
                WeightProcessingException.ProcessingMethodInvalid,
                "The weight-processing method must be Standard, Floor, or Ceiling."),
        };

        var processedScaledWeight = units * quantum;

        return new WeightProcessingResult(
            rawWeightKg!,
            FormatWeight(processedScaledWeight, StorageDecimalPlaces),
            FormatWeight(processedScaledWeight, policy.DecimalPlaces),
            policy.DecimalPlaces,
            policy.Method);
    }

    private static BigInteger ParseRawWeight(string? rawWeightKg)
    {
        if (string.IsNullOrWhiteSpace(rawWeightKg))
        {
            throw new WeightProcessingException(
                WeightProcessingException.RawRequired,
                "Raw weight is required.");
        }

        if (rawWeightKg[0] == '-')
        {
            throw new WeightProcessingException(
                WeightProcessingException.RawNegative,
                "Raw weight cannot be negative.");
        }

        var decimalSeparatorIndex = rawWeightKg.IndexOf('.');
        if (decimalSeparatorIndex == 0 ||
            decimalSeparatorIndex == rawWeightKg.Length - 1 ||
            decimalSeparatorIndex != rawWeightKg.LastIndexOf('.'))
        {
            throw InvalidRawWeight();
        }

        for (var index = 0; index < rawWeightKg.Length; index++)
        {
            if (index == decimalSeparatorIndex)
            {
                continue;
            }

            if (rawWeightKg[index] is < '0' or > '9')
            {
                throw InvalidRawWeight();
            }
        }

        var integerDigits = decimalSeparatorIndex < 0
            ? rawWeightKg.Length
            : decimalSeparatorIndex;
        var fractionDigits = decimalSeparatorIndex < 0
            ? 0
            : rawWeightKg.Length - decimalSeparatorIndex - 1;

        if (integerDigits > MaximumIntegerDigits)
        {
            throw new WeightProcessingException(
                WeightProcessingException.RawIntegerDigitsExceeded,
                $"Raw weight cannot have more than {MaximumIntegerDigits} integer digits.");
        }

        if (fractionDigits > MaximumFractionDigits)
        {
            throw new WeightProcessingException(
                WeightProcessingException.RawFractionDigitsExceeded,
                $"Raw weight cannot have more than {MaximumFractionDigits} fractional digits.");
        }

        var integerPart = BigInteger.Parse(
            rawWeightKg.AsSpan(0, integerDigits),
            NumberStyles.None,
            CultureInfo.InvariantCulture);
        var fractionalPart = fractionDigits == 0
            ? BigInteger.Zero
            : BigInteger.Parse(
                rawWeightKg.AsSpan(
                    decimalSeparatorIndex + 1,
                    fractionDigits),
                NumberStyles.None,
                CultureInfo.InvariantCulture) *
              BigInteger.Pow(
                  10,
                  StorageDecimalPlaces - fractionDigits);

        return integerPart * StorageScale + fractionalPart;
    }

    private static string FormatWeight(
        BigInteger scaledWeight,
        int decimalPlaces)
    {
        var integerPart = BigInteger.DivRem(
            scaledWeight,
            StorageScale,
            out var sixDigitFraction);
        var displayDivisor = BigInteger.Pow(
            10,
            StorageDecimalPlaces - decimalPlaces);
        var displayedFraction = sixDigitFraction / displayDivisor;

        return string.Concat(
            integerPart.ToString(CultureInfo.InvariantCulture),
            ".",
            displayedFraction.ToString(
                $"D{decimalPlaces}",
                CultureInfo.InvariantCulture));
    }

    private static WeightProcessingException InvalidRawWeight()
    {
        return new WeightProcessingException(
            WeightProcessingException.RawInvalid,
            "Raw weight must use unsigned decimal notation with a dot separator.");
    }
}

File: apps/mobile/lib/core/measurements/weight_processor.dart
import 'weight_processing_exception.dart';
import 'weight_processing_method.dart';
import 'weight_processing_policy.dart';
import 'weight_processing_result.dart';

abstract final class WeightProcessor {
  static const _storageDecimalPlaces = 6;
  static const _maximumIntegerDigits = 14;
  static const _maximumFractionDigits = 6;
  static final BigInt _storageScale = BigInt.from(
    10,
  ).pow(_storageDecimalPlaces);

  static WeightProcessingResult process({
    required String rawWeightKg,
    required int decimalPlaces,
    required String method,
  }) {
    return processWithPolicy(
      rawWeightKg: rawWeightKg,
      policy: WeightProcessingPolicy.parse(
        decimalPlaces: decimalPlaces,
        method: method,
      ),
    );
  }

  static WeightProcessingResult processWithPolicy({
    required String rawWeightKg,
    required WeightProcessingPolicy policy,
  }) {
    final rawScaledWeight = _parseRawWeight(rawWeightKg);
    final quantum = BigInt.from(
      10,
    ).pow(_storageDecimalPlaces - policy.decimalPlaces);
    var units = rawScaledWeight ~/ quantum;
    final remainder = rawScaledWeight.remainder(quantum);

    switch (policy.method) {
      case WeightProcessingMethod.standard:
        if (remainder * BigInt.two >= quantum) {
          units += BigInt.one;
        }
      case WeightProcessingMethod.floor:
        break;
      case WeightProcessingMethod.ceiling:
        if (remainder != BigInt.zero) {
          units += BigInt.one;
        }
    }

    final processedScaledWeight = units * quantum;

    return WeightProcessingResult(
      rawWeightKg: rawWeightKg,
      processedWeightKg: _formatWeight(
        processedScaledWeight,
        _storageDecimalPlaces,
      ),
      displayWeightKg: _formatWeight(
        processedScaledWeight,
        policy.decimalPlaces,
      ),
      decimalPlaces: policy.decimalPlaces,
      method: policy.method,
    );
  }

  static BigInt _parseRawWeight(String rawWeightKg) {
    if (rawWeightKg.trim().isEmpty) {
      throw const WeightProcessingException(
        WeightProcessingException.rawRequired,
        'Raw weight is required.',
      );
    }

    if (rawWeightKg.startsWith('-')) {
      throw const WeightProcessingException(
        WeightProcessingException.rawNegative,
        'Raw weight cannot be negative.',
      );
    }

    final decimalSeparatorIndex = rawWeightKg.indexOf('.');
    if (decimalSeparatorIndex == 0 ||
        decimalSeparatorIndex == rawWeightKg.length - 1 ||
        decimalSeparatorIndex != rawWeightKg.lastIndexOf('.')) {
      throw _invalidRawWeight();
    }

    for (var index = 0; index < rawWeightKg.length; index++) {
      if (index == decimalSeparatorIndex) {
        continue;
      }

      final codeUnit = rawWeightKg.codeUnitAt(index);
      if (codeUnit < 0x30 || codeUnit > 0x39) {
        throw _invalidRawWeight();
      }
    }

    final integerDigits = decimalSeparatorIndex < 0
        ? rawWeightKg.length
        : decimalSeparatorIndex;
    final fractionDigits = decimalSeparatorIndex < 0
        ? 0
        : rawWeightKg.length - decimalSeparatorIndex - 1;

    if (integerDigits > _maximumIntegerDigits) {
      throw const WeightProcessingException(
        WeightProcessingException.rawIntegerDigitsExceeded,
        'Raw weight cannot have more than 14 integer digits.',
      );
    }

    if (fractionDigits > _maximumFractionDigits) {
      throw const WeightProcessingException(
        WeightProcessingException.rawFractionDigitsExceeded,
        'Raw weight cannot have more than 6 fractional digits.',
      );
    }

    final integerPart = BigInt.parse(rawWeightKg.substring(0, integerDigits));
    final fractionalPart = fractionDigits == 0
        ? BigInt.zero
        : BigInt.parse(rawWeightKg.substring(decimalSeparatorIndex + 1)) *
              BigInt.from(10).pow(_storageDecimalPlaces - fractionDigits);

    return integerPart * _storageScale + fractionalPart;
  }

  static String _formatWeight(BigInt scaledWeight, int decimalPlaces) {
    final integerPart = scaledWeight ~/ _storageScale;
    final sixDigitFraction = scaledWeight.remainder(_storageScale);
    final displayDivisor = BigInt.from(
      10,
    ).pow(_storageDecimalPlaces - decimalPlaces);
    final displayedFraction = sixDigitFraction ~/ displayDivisor;

    return '${integerPart.toString()}.'
        '${displayedFraction.toString().padLeft(decimalPlaces, '0')}';
  }

  static WeightProcessingException _invalidRawWeight() {
    return const WeightProcessingException(
      WeightProcessingException.rawInvalid,
      'Raw weight must use unsigned decimal notation with a dot separator.',
    );
  }
}
